types/prefs/prefs.go

```diff
@@ -0,0 +1,179 @@
+type serializable[T any] struct {
+	// Value is an optional preference value that is set when the preference is
+	// configured by the user or managed by an admin.
+	Value opt.Value[T] `json:",omitzero"`
+	// Default is the default preference value to be used
+	// when the preference has not been configured.
+	Default T `json:",omitzero"`
+	// Metadata is any additional type-agnostic preference metadata to be serialized.
+	Metadata metadata `json:",inline"`
+}
+
+// preference is an embeddable type that provides a common implementation for
+// concrete preference types, such as [Item], [List], [Map], [StructList] and [StructMap].
+type preference[T any] struct {
+	s serializable[T]
+}
+
+// preferenceOf returns a preference with the specified value and/or [Options].
+func preferenceOf[T any](v opt.Value[T], opts ...Options) preference[T] {
+	var m metadata
+	for _, o := range opts {
+		o(&m)
+	}
+	return preference[T]{serializable[T]{Value: v, Metadata: m}}
+}
+
+// IsSet reports whether p has a value set.
+func (p preference[T]) IsSet() bool {
+	return p.s.Value.IsSet()
+}
+
+// Value returns the value of p if the preference has a value set.
+// Otherwise, it returns its default value.
+func (p preference[T]) Value() T {
+	val, _ := p.ValueOk()
+	return val
+}
+
+// ValueOk returns the value of p and true if the preference has a value set.
+// Otherwise, it returns its default value and false.
+func (p preference[T]) ValueOk() (val T, ok bool) {
+	if val, ok = p.s.Value.GetOk(); ok {
+		return val, true
+	}
+	return p.DefaultValue(), false
+}
+
+// SetValue configures the preference with the specified value.
+// It fails and returns [ErrManaged] if p is a managed preference,
+// and [ErrReadOnly] if p is a read-only preference.
+func (p *preference[T]) SetValue(val T) error {
+	switch {
+	case p.s.Metadata.Managed:
+		return ErrManaged
+	case p.s.Metadata.ReadOnly:
+		return ErrReadOnly
+	default:
+		p.s.Value.Set(val)
+		return nil
+	}
+}
+
+// ClearValue resets the preference to an unconfigured state.
+// It fails and returns [ErrManaged] if p is a managed preference,
+// and [ErrReadOnly] if p is a read-only preference.
+func (p *preference[T]) ClearValue() error {
+	switch {
+	case p.s.Metadata.Managed:
+		return ErrManaged
+	case p.s.Metadata.ReadOnly:
+		return ErrReadOnly
+	default:
+		p.s.Value.Clear()
+		return nil
+	}
+}
+
+// DefaultValue returns the default value of p.
+func (p preference[T]) DefaultValue() T {
+	return p.s.Default
+}
+
+// SetDefaultValue sets the default value of p.
+func (p *preference[T]) SetDefaultValue(def T) {
+	p.s.Default = def
+}
+
+// IsManaged reports whether p is managed via MDM, Group Policy, or similar means.
+func (p preference[T]) IsManaged() bool {
+	return p.s.Metadata.Managed
+}
+
+// SetManagedValue configures the preference with the specified value
+// and marks the preference as managed.
+func (p *preference[T]) SetManagedValue(val T) {
+	p.s.Value.Set(val)
+	p.s.Metadata.Managed = true
+}
```

Comment:
nit: maybe just `SetManaged` that sets the metadata flag only? otherwise it's inconsistent with `SetReadOnly`.

Reply:
I think the main reason I chose to implement it this way was that setting a managed value is logically atomic: one should not be able to set the `Managed` flag without also providing a value, or configure a preference with an admin-provided value without also marking the preference as `Managed`. This approach would also make it easier to:
- track managed value origin (e.g. local config, MDM or control plane policy);
- revert the preference to the user's choice when it is no longer managed.

Reply:
We decided to keep it as is.